New Resource: `azurerm_virtual_machine_implicit_data_disk_from_source`

ms-zhenhua opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

**Note:** ideally, this feature should be supported by adding `data_disks` in `azurerm_[linux|windows]_virtual_machine`. However, due to [technical limitations](https://github.com/hashicorp/terraform-provider-azurerm/pull/10466#issuecomment-828194670), a new resource is defined to support this feature. 

This new resource is used to support a new virtual machine feature named **implicit disk creation**. Being different from the existing resource `azurerm_virtual_machine_data_disk_attachment`, which attaches an existing managed disk to the VM, this new resource creates a new data disk when updating a VM and allows the system to place the this disk closer to the parent VM to provide better latency and align the fault domain of the disk with the parent VM for providing better reliability.

This resource is similar with `azurerm_virtual_machine_data_disk_attachment` so that they share the same ID format. However, since this new resource uses some unique new properties, it is defined as a separate resource. These two resources are distinguished by different values of the `create_option`, which are added into the `import` code.



## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/76987228/f1617c0e-0b25-4fd5-96c1-73ac1786e60c)

![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/76987228/263f9569-8cf3-4cd2-95e0-23ae0e793283)


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_virtual_machine_implicit_data_disk_from_source` - a new resource to support implicit disk creation feature for virtual machines


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes #0000


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

tombuildsstuff commented on internal/services/compute/virtual_machine_implicit_data_disk_resource.go:
```diff
@@ -0,0 +1,462 @@
+							return fmt.Errorf("deleting %s: %+v", id, err)
+						}
+
+						// delete data disk if delete_option is set to delete
+						if toBeDeletedDisk != nil && pointer.From(toBeDeletedDisk.DeleteOption) == virtualmachines.DiskDeleteOptionTypesDelete &&
+							toBeDeletedDisk.ManagedDisk != nil && toBeDeletedDisk.ManagedDisk.Id != nil {
+							diskClient := metadata.Client.Compute.DisksClient
+							diskId, err := commonids.ParseManagedDiskID(*toBeDeletedDisk.ManagedDisk.Id)
+							if err != nil {
+								return err
+							}
+
+							err = diskClient.DeleteThenPoll(ctx, *diskId)
+							if err != nil {
+								return fmt.Errorf("deleting Managed Disk %s: %+v", *diskId, err)
+							}
+						}
```
This is an API setting - why are we deleting the disk here?

ms-zhenhua replied:
Because this disk is created by the backend so that it's not managed by any TF resource such as `azurerm_managed_disk`. If we do not delete it explicitly here, it will only be removed from the VM configuration, but the disk itself still exists. When VM is deleted, this disk will not be deleted together because it's already been removed from the VM configuration. Since the disk still exists in the resource group, the resource group cannot be deleted unless `prevent_deletion_if_contains_resources=false`. Due to this reason, I choose to explicitly delete the disk when `DeleteOption=true`

mbfrahry replied:
Hmmm @ms-zhenhua, what're your thoughts on deleting the disk when it's unhooked from the vm and not worrying about `delete_option`? This resource means that Terraform is creating the disk and I would expect Terraform to also delete the disk

Are there use cases when we want to keep the disk around?

ms-zhenhua replied:
Hi @mbfrahry, when we create this resource, what we really do is to add a data disk configuration to the existing VM, which will trigger the Azure backend to create an implicit data disk for the VM. However, when we delete this resource, what we really do is just remove the data disk configuration from the VM, but the physical data disk is only detached from the VM and still exists in the resource group. That's why I use the delete option to manually delete this data disk, otherwise, the resource group cannot be deleted normally.

tombuildsstuff commented on internal/services/compute/virtual_machine_implicit_data_disk_resource.go:
```diff
@@ -0,0 +1,462 @@
+				WriteAcceleratorEnabled: pointer.To(config.WriteAcceleratorEnabled),
+			}
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					if profile := props.StorageProfile; profile != nil {
+						if dataDisks := profile.DataDisks; dataDisks != nil {
```
there's ways to provision a VM without a StorageProfile and/or DataDisks (e.g. when creating from certain types of image) - as such we'd end up doing nothing in those cases here - we'd need to update this to account for that

ms-zhenhua replied:
code updated. Thanks.

tombuildsstuff commented on internal/services/compute/virtual_machine_implicit_data_disk_resource.go:
```diff
@@ -0,0 +1,462 @@
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ValidateFunc: validation.StringInSlice([]string{
+				string(virtualmachines.CachingTypesNone),
```
See the `None` topic in the Contributor Guide

ms-zhenhua replied:
code updated. Thanks.

tombuildsstuff commented on internal/services/compute/virtual_machine_implicit_data_disk_resource.go:
```diff
@@ -0,0 +1,462 @@
+	_ sdk.ResourceWithCustomImporter = VirtualMachineImplicitDataDiskResource{}
+)
+
+type VirtualMachineImplicitDataDiskResource struct{}
```
Per the PR description:

> this new resource creates a new data disk when updating a VM and allows the system to place the this disk closer to the parent VM to provide better latency and align the fault domain of the disk with the parent VM for providing better reliability.

Can you provide a source showing the performance benefits of this? Last I read using a Zone/Proximity Placement Group should give the same benefit here - and from what I understood was replacing the Fault Domains concept (which is associated with Availability Sets, which I believe aren't recommended anymore?)?

Based on this issue (https://github.com/Azure/azure-cli/issues/28686) - since this exists only for copies of a Disk from a Snapshot, rather than a net-new Disk - I'm concerned that the resource name is going to be misleading too?

tombuildsstuff reviewed (changes requested):
hey @ms-zhenhua 

I've taken a look through and left some comments inline - but as it stands I suspect the name of this Resource is going to be a source of confusion in its current form.

Since we're creating a Data Disk from a Snapshot and then attaching that - perhaps a better name for this would be `azurerm_virtual_machine_implicit_data_disk_from_source`?

Thanks!

ms-zhenhua commented:
Hi @tombuildsstuff, thank you for the review. I have updated the code and comments. Could you kindly have another review?

mbfrahry commented on internal/services/compute/virtual_machine_implicit_data_disk_from_source.go:
```diff
@@ -0,0 +1,476 @@
+package compute
```
This filename wants to be `virtual_machine_implicit_data_disk_from_source_resource.go`

ms-zhenhua replied:
updated. Thanks.

mbfrahry commented on internal/services/compute/virtual_machine_implicit_data_disk_from_source_test.go:
```diff
@@ -0,0 +1,992 @@
+package compute_test
```
And this filename wants to be `virtual_machine_implicit_data_disk_from_source_resource_test.go`

ms-zhenhua replied:
updated. Thanks.

mbfrahry commented on internal/services/compute/virtual_machine_data_disk_attachment_resource.go:
```diff
@@ -29,9 +30,51 @@ func resourceVirtualMachineDataDiskAttachment() *pluginsdk.Resource {
+				return nil, fmt.Errorf("Data Disk %s was not found", *id)
+			}
+
+			if disk.CreateOption != virtualmachines.DiskCreateOptionTypesAttach && disk.CreateOption != virtualmachines.DiskCreateOptionTypesEmpty {
```
Is it possible to make a test for this?

mbfrahry commented on internal/services/compute/virtual_machine_data_disk_attachment_resource.go:
```diff
@@ -154,6 +197,15 @@ func resourceVirtualMachineDataDiskAttachmentCreateUpdate(d *pluginsdk.ResourceD
 		WriteAcceleratorEnabled: pointer.To(writeAcceleratorEnabled),
 	}
 
+	// there are ways to provision a VM without a StorageProfile and/or DataDisks
```
This should also include a test to confirm that the existing VM won't have any diffs when a disk is attached this way

ms-zhenhua replied:
1. I have re-run all existing testcases which are passed and attached in the [description](https://github.com/hashicorp/terraform-provider-azurerm/pull/25537#issue-2232455538).
2. Though this [comment ](https://github.com/hashicorp/terraform-provider-azurerm/pull/25537#discussion_r1557090305) said there are ways to provision a VM without a StorageProfile and/or DataDisks with certain types of images, I do not really know which image will trigger such scenario. Kindly let me know if you know the image type and I will add the testcase. However, this part of code will not cause any diff because this code only helps initialize the storage and disk when they are `nil`.

mbfrahry commented on internal/services/compute/virtual_machine_implicit_data_disk_from_source.go:
```diff
@@ -0,0 +1,476 @@
+						}
+
+						// delete data disk if delete_option is set to delete
+						if toBeDeletedDisk != nil && pointer.From(toBeDeletedDisk.DeleteOption) == virtualmachines.DiskDeleteOptionTypesDelete &&
```
Because Terraform is creating this resource, I believe we can make the assumption that Terraform can delete this resource? If there is value in keeping the disk around, we can add a new variable to the  `features` block that will hang on to the disk instead of deleting it. Or vice versa if the default should be to keep the disk around instead

ms-zhenhua replied:
This resource is a little special. It manages the [data disk](https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update?view=rest-compute-2024-03-01&tabs=HTTP#datadisk) of a VM such as `azurerm_virtual_machine_data_disk_attachment`. When creating this resource, what is really done is to add a data disk configuration to the existing VM, which will trigger the Azure service to create a new data disk for the VM. When deleting this resource, what is really done is to remove the data disk configuration from the existing VM. However, the physical data disk will not be removed from the resource group, which may cause the resource group cannot be deleted later. That's why I add this ugly delete option. According to [your comment](https://github.com/hashicorp/terraform-provider-azurerm/pull/25537#discussion_r1607415824),  do you mean we can hang on to the disk by default and should add a new variable to the provider [features block](https://github.com/hashicorp/terraform-provider-azurerm/blob/343bf3650e0da580c30f8b15d843f7288c4c8314/internal/provider/provider.go#L320) to control whether to delete the physical data disk when deleting this resource?

mbfrahry replied:
This is tough because from the users perspective, Terraform is creating a disk (I do see that Azure is creating the disk from what we are sending to the api but the user will see that it is Terraform that is creating the disk) so running destroy should delete the disk that was created. 

My thought is to have the default be to delete this disk and then in a features block, we add an option to detach the disk instead of deleting it.

mbfrahry reviewed (changes requested):
Hey @ms-zhenhua, I've left some more comments, I'm still hung up on the Delete method and would like some further discussion on that piece

ms-zhenhua commented:
Hi @mbfrahry , thank you for your review. I have updated the code and replied your comments. Please help take another review and leave your questions and comments. Thanks.

ms-zhenhua commented:
Hi @mbfrahry, thank you for your advice. I removed `DeleteOption` and made `Delete` as the default operation. I added a variable in the features block to control whether to detach the data disk when destroying the resource. Please take another review. Thanks.

mbfrahry commented on website/docs/guides/features-block.html.markdown:
```diff
@@ -240,6 +241,8 @@ The `template_deployment` block supports the following:
 
 The `virtual_machine` block supports the following:
 
+* `detach_implicit_data_disk_on_deletion` - (Optional) Should the `azurerm_virtual_machine_implicit_data_disk_from_source` resource detach the implicit data disk of the Virtual Machine when the `azurerm_virtual_machine_implicit_data_disk_from_source` resource is destroyed? Defaults to `false`.
```
```suggestion
* `detach_implicit_data_disk_on_deletion` - (Optional) Should we detach the `azurerm_virtual_machine_implicit_data_disk_from_source` from the virtual machine instead of destroying it? Defaults to `false`.
```

mbfrahry reviewed (approved):
LGTM! Thanks for working through the reviews @ms-zhenhua !
